### ISSUE TITLE
[HTTP/3] Stress test failures for GET Aborted scenario

### ISSUE THREAD

**Issue by CarnaViire** [dotnet/runtime#72619]
Most of the failures (~399) are about scenario not getting an exception it expected to get from the server:

```
client_1  | System.Exception: Completed unexpectedly
client_1  |    at HttpStress.ClientOperations.<>c.<<get_Operations>b__1_4>d.MoveNext() in /app/ClientOperations.cs:line 313
client_1  | --- End of stack trace from previous location ---
client_1  |    at HttpStress.StressClient.<>c__DisplayClass17_0.<<StartCore>g__RunWorker|0>d.MoveNext() in /app/StressClient.cs:line 224
```

A small part of failures (~5) is:

```
client_1  | System.Net.Http.HttpRequestException: The server returned an invalid or unrecognized response.
client_1  |    at System.Net.Http.Http3RequestStream.ReadResponseAsync(CancellationToken cancellationToken) in /_/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:line 337
client_1  |    at System.Net.Http.Http3RequestStream.SendAsync(CancellationToken cancellationToken) in /_/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:line 304
client_1  |    at System.Net.Http.Http3RequestStream.SendAsync(CancellationToken cancellationToken) in /_/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:line 304
client_1  |    at System.Net.Http.Http3Connection.SendAsync(HttpRequestMessage request, Int64 queueStartingTimestamp, CancellationToken cancellationToken) in /_/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs:line 245
client_1  |    at System.Net.Http.HttpConnectionPool.TrySendUsingHttp3Async(HttpRequestMessage request, CancellationToken cancellationToken) in /_/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:line 974
client_1  |    at System.Net.Http.HttpConnectionPool.SendWithVersionDetectionAndRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken) in /_/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:line 1116
client_1  |    at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken) in /_/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/RedirectHandler.cs:line 85
client_1  |    at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken) in /_/src/libraries/System.Net.Http/src/System/Net/Http/HttpClient.cs:line 558
client_1  |    at HttpStress.RequestContext.SendAsync(HttpRequestMessage request, HttpCompletionOption httpCompletion, Nullable`1 token) in /app/ClientOperations.cs:line 108
client_1  |    at HttpStress.ClientOperations.<>c.<<get_Operations>b__1_4>d.MoveNext() in /app/ClientOperations.cs:line 313
client_1  | --- End of stack trace from previous location ---
client_1  |    at HttpStress.StressClient.<>c__DisplayClass17_0.<<StartCore>g__RunWorker|0>d.MoveNext() in /app/StressClient.cs:line 224
```

Last Main run: https://dev.azure.com/dnceng/public/_build/results?buildId=1890195&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=1451f5f3-0108-5a08-5b92-e984b2a85bbd&l=1632
Most recent run (on #72601) https://dev.azure.com/dnceng/public/_build/results?buildId=1895244&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=1451f5f3-0108-5a08-5b92-e984b2a85bbd&l=1596

**Comment by CarnaViire** [comment]
"Completed unexpectedly" still happens: [Run #20220804.2](https://dev.azure.com/dnceng/public/_build/results?buildId=1922992&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=1451f5f3-0108-5a08-5b92-e984b2a85bbd&l=1127) from yesterday (04.08)

**Comment by ManickaP** [comment]
It started happening again between 3.8. and 4.8. There was a change in ASP 7.0.0-rc.1.22402.14 --> 7.0.0-rc.1.22403.3 (I don't see any changes in S.N.Quic that might have caused this).
I see that in https://github.com/dotnet/aspnetcore/pull/43011 `Abort` and `DisposeAsync` on the server side stream were moved outside of `_shutdownLock`. @JamesNK is it possible now that `DisposeAsync` races with the `Abort`?

**Comment by JamesNK** [comment]
What do you mean by races? What do you expect to happen when a stream is aborted while `QuicStream.DisposeAsync` is being called?

Previously ASP.NET Core was calling `QuicStream.Dispose`. That's method just calls `QuicStream.DisposeAsync` but with sync-over-async, which we obviously don't want.

I can't repo getting "The server returned an invalid or unrecognized response." in gRPC benchmarks. However, none of them abort.

**Comment by ManickaP** [comment]
My theory is that  `HttpContext.Abort` finished before `QuicStream.Abort` was actually called (that there's a disconnect between those two operations) and it races with `QuicStream.DisposeAsync`. But please correct me if I'm wrong. There're several layers of code in between and I might be missing something.

For completeness, our server code:
https://github.com/dotnet/runtime/blob/c99545ca6458d0ff0c7facef8ec407e7d2bdf896/src/libraries/System.Net.Http/tests/StressTests/HttpStress/StressServer.cs#L241-L247
Which should result in aborted stream on the client side.

From user perspective, calling `Abort` on `HttpContext` should not end before you actually call `Abort` on `QuicStream`.

**Comment by JamesNK** [comment]
`HttpContext.Abort` calls `QuicStream.Abort` inline (i.e. doesn't dispatch to a different thread)

![image](https://user-images.githubusercontent.com/303201/183786815-9a097932-50af-481c-9e79-6f9701cd0072.png)

DisposeAsync is always called after the request user code is complete. The only way they could conflict is by someone calling `HttpContext.Abort` on a different thread while the request is being cleaned up, and that's completely unsupported.

**Comment by JamesNK** [comment]
I wrote a test to see if I could duplicate it - https://github.com/dotnet/aspnetcore/pull/43183. Runs fine with 1000 requests (at least on my machine)

**Comment by karelz** [comment]
Based on discussion with @ManickaP - this could be caused by QuicStream on server sending success (incl. FIN flag) when server code aborted the stream and did not send full response yet. As a result, the response on the wire appears to be complete and successful even when it is incomplete from server code point of view.

**Comment by CarnaViire** [comment]
Could it happen that TrySetException would for some reason unexpectedly return false here and we will end up never issuing an Abort? And subsequent Dispose will end up closing the stream gracefully...

https://github.com/dotnet/runtime/blob/8a4de2ebe5bd13f3622da314e6c137f8b028e3ee/src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs#L409-L412

Note that abort for read side currently will happen regardless of TrySetException result, though it might be a result of some copy-paste error, because the code adds the flag twice:
https://github.com/dotnet/runtime/blob/8a4de2ebe5bd13f3622da314e6c137f8b028e3ee/src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs#L401-L405

**Comment by JamesNK** [comment]
Kestrel could move the abort back into a lock if it's needed.

My question is, should QuicStream be making these checks? i.e. CompleteWrites no-ops if an abort has happened. Or throws an error. It doesn't seem like a good user experience to require all calling code to have locks when QuicStream could do it itself. Is QuicStream intended to be thread-safe?

**Comment by rzikm** [comment]
> Is QuicStream intended to be thread-safe?
I believe the answer is no

**Comment by ManickaP** [comment]
> i.e. CompleteWrites no-ops if an abort has happened.

That's what happens internally in msquic, we do not prevent the call though. But if `CompleteWrites` arrives after `Abort` it won't do anything.

And I don't think it's even in `QuicStream` ability to prevent this race. If there's possibility of having `CompleteWrites` or `WriteAsync(completeWrites: true)` run in parallel with `Abort`, no amount of locking in `QuicStream` will make 100% sure that `Abort` gets called first. 

If I'm reading the code of `AbortCore` right:
https://github.com/dotnet/aspnetcore/blob/51ccce6d743dddf995b3661f07b0e1990b68dc7d/src/Servers/Kestrel/Core/src/Internal/Http3/Http3Stream.cs#L145-L184
It first completes the `PipeWriter` before calling `QuicStream.Abort`:
https://github.com/dotnet/aspnetcore/blob/51ccce6d743dddf995b3661f07b0e1990b68dc7d/src/Servers/Kestrel/Core/src/Internal/Http3/Http3Stream.cs#L170-L172

And the `PipeWriter` completion will asynchronously attempt to "complete the writes" gracefully.

@JamesNK do you want me to file an issue in your repo? Transfer this one?

**Comment by JamesNK** [comment]
Log an issue please. Leave this here to confirm that changing asp.net core resolves the stress failures.

**Comment by JamesNK** [comment]
Make a change: https://github.com/dotnet/aspnetcore/pull/43630

Does it have any impact?

**Comment by karelz** [comment]
@CarnaViire would you be able to check it?

**Comment by CarnaViire** [comment]
I confirm that there are no such failures anymore, neither locally nor in CI.
HTTP/3 stress now is green in CI on both main and release/7.0 🥳
Thanks @JamesNK!
